src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/PointerLocation.java

```diff
@@ -186,24 +213,66 @@ public String toString() {
   }
 
   public void print() {
-    printOn(System.out);
+      printOn(System.out, true, true);
+  }
+
+  public void print(boolean printAddress, boolean verbose) {
+    printOn(System.out, printAddress, verbose);
   }
 
   public void printOn(PrintStream tty) {
-    tty.print("Address ");
-    if (addr == null) {
-      tty.print("0x0");
-    } else {
-      tty.print(addr.toString());
+    printOn(tty, true, true);
+  }
+
+  public void printOn(PrintStream tty, boolean printAddress, boolean verbose) {
+    if (printAddress) {
+      tty.print("Address ");
+      if (addr == null) {
+        tty.print("0x0");
+      } else {
+        tty.print(addr.toString());
+      }
+      tty.print(": ");
     }
-    tty.print(": ");
-    if (isInHeap()) {
+    if (isMetadata()) {
+      metadata.printValueOn(tty); // does not include "\n"
+      tty.println();
+    } else if (isCtype()) {
+      tty.println("Is of type " + ctype.getName());
+    } else if (isInJavaStack()) {
+        if (verbose) {
+            tty.format("In java stack [%s,%s,%s] for thread %s:\n   ",
+                       stackThread.getStackBase(), stackThread.lastSPDbg(),
+                       stackThread.getStackBase().addOffsetTo(-stackThread.getStackSize()),
+                       stackThread);
```

Comment:
When we print a JavaThread, in the verbose block,
the final argument to tty.format in line 247, I wonder what that prints?

We then call printThreadInfoOn() which will first print the quoted thread name,
so maybe we don't need that item.
Or maybe we want the JavaThread.toString()?

Reply:
`stackThread.toString()` ends up in `VMObject.toString()`:

```
  public String toString() {
    return getClass().getName() + "@" + addr;
  }
```
And here's an example output:
```
hsdb> + findpc 0x0000152f45df6000
Address 0x0000152f45df6000: In java stack [0x0000152f45df8000,0x0000152f45df6580,0x0000152f45cf7000] for thread sun.jvm.hotspot.runtime.JavaThread@0x0000152f3c026f70:
   "main" #1 prio=5 tid=0x0000152f3c026f70 nid=0x308e waiting on condition [0x0000152f45df6000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
   JavaThread state: _thread_blocked
```
So I think the `stackThread` argument is doing what was intended, and there is no duplication in the output.

Reply:
Great, thanks.